### PR TITLE
Emit empty primary constructor

### LINKED ISSUE #1093: Empty primary constructor not generated properly
Hello! 

I am using Kotlinpoet v1.8.0 and I am trying to generate an empty primary constructor for a class `TypeSpec`.

```kotlin
builderSpec.primaryConstructor(FunSpec.constructorBuilder().build())
```

The expected result here would be something like:

```kotlin
class Foo() { ... }
```

whereas the actual result is:

```kotlin
class Foo { ... }
```

This breaks the functionality when we are trying to add a second constructor that uses `callThisConstructor()` since there is no primary empty constructor to delegate to.

### PR DESCRIPTION
**Pull request by pavlospt**
Fixes #1093

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 1]
```diff
@@ -186,17 +186,21 @@ public class TypeSpec private constructor(
...
+                  codeWriter, setOf(PUBLIC), withInitializer = false, inline = true,
+                  inlineAnnotations = false
+                )
+                param.emitDefaultValue(codeWriter)
+              } else {
+                param.emit(codeWriter, emitKdoc = true, inlineAnnotations = false)
+              }
             }
+          } else {
+            codeWriter.emit("()")
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt:203]: `parameters.emit()` already encodes this behavior if `forceParensOnEmpty` is true, so I suggest using that. It also looks like this is the only call site that passes a value for `forceParensOnEmpty` so we can explore deleting that parameter altogether and just making the method always emit brackets.

Comment (pavlospt) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Update with your comments @Egorand 😃

**Review by Egorand — APPROVED** [review]
Amazing, thanks!

**Comment by Egorand** [comment]
@pavlospt ready to merge once spotless issues are fixed - just run `spotlessApply` and push!

**Comment by pavlospt** [comment]
> @pavlospt ready to merge once spotless issues are fixed - just run `spotlessApply` and push!

Ready @Egorand !

**Comment by Egorand** [comment]
@pavlospt looks like we need more test updates here

**Comment by pavlospt** [comment]
> @pavlospt looks like we need more test updates here

Yeap I am on it!

**Comment by pavlospt** [comment]
@Egorand hopefully fixed. Locally everything passed !

**Comment by ZacSweers** [comment]
Could we make this a bit more intelligent? I.e. rather than always emit, can we only emit if any constructors have super calls?

**Comment by pavlospt** [comment]
> Could we make this a bit more intelligent? I.e. rather than always emit, can we only emit if any constructors have super calls?

Since this code is currently emitted based on the constructor parameters I can not seem to find a quick and obvious solution on that. Maybe with a bigger refactoring this could be doable!

**Comment by Egorand** [comment]
@ZacSweers we'll only emit when explicitly given a `primaryConstructor(FunSpec.constructorBuilder().build())`, which I think is correct - if the use site doesn't want the empty constructor it should omit passing the `FunSpec`.

**Pull request merged**